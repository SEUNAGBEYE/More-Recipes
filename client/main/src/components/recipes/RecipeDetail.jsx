import React, { Component } from 'react';
import Loader from 'react-loader';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import checkAuth from '../../../utils/CheckAuth';
import CreateReview from '../recipes/CreateReview';
import Review from '../recipes/Review';
import DeleteModal from './DeleteModal';
import EditModal from './EditModal';
import { getRecipe,
  reviewRecipe,
  toggleFavouriteRecipe,
  toggleThumbsDownRecipe,
  toggleThumbsUpRecipe,
  getRecipeReviews,
  deleteRecipe,
  editRecipe } from '../../actions/Recipes';
import RecipeCardActions from '../recipes/RecipeCardActions';
import Exclamation from './Exclamation';


/**
 * @export
 * @class RecipeDetail
 * @extends {Component}
 */
export class RecipeDetail extends Component {
  /**
   * @description - Creates an instance of RecipeDetail.
   *
   * @method constructor
   *
   * @param {Object} props
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  constructor(props) {
    super(props);
    this.state = {
      reviewsPage: 1,
      reviewBody: '',
      reviewsLoaded: true,
      loading: true
    };
    this.toggleFavouriteRecipe = this.toggleFavouriteRecipe.bind(this);
    this.toggleThumbsDownRecipe = this.toggleThumbsDownRecipe.bind(this);
    this.toggleThumbsUpRecipe = this.toggleThumbsUpRecipe.bind(this);
    this.viewMoreReviews = this.viewMoreReviews.bind(this);
    this.onChange = this.onChange.bind(this);
    this.reviewRecipe = this.reviewRecipe.bind(this);
    this.deleteRecipe = this.deleteRecipe.bind(this);
  }

  /**
   * @description - Get Recipe Detail Data
   *
   * @method componentDidMount
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  componentDidMount() {
    this.props.getRecipe(this.props.match.params.id).then(() => {
      this.setState({ loading: false });
    });
  }

  /**
   * @description - Change state values
   *
   * @method onChange
   *
   * @param {Object} event
   *
   * @returns {void}
   * @memberof RecipeDetail
   */
  onChange(event) {
    checkAuth(this.props.isAuthenticated, this.props.history);
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  /**
   * @description - Delete a recipe
   *
   * @method deleteRecipe
   *
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  deleteRecipe(event) {
    const { id } = event.target;
    this.props.deleteRecipe(id)
      .then(() => this.props.history.push('/my_recipes'));
  }


  /**
   * @description - Add review to a recipe
   *
   * @method reviewRecipe
   *
   * @param {Object} event
   *
   * @returns {void}
   * @memberof RecipeDetail
   */
  reviewRecipe(event) {
    event.preventDefault();
    const { isAuthenticated, history, recipe } = this.props;
    checkAuth(isAuthenticated, history);
    this.props.reviewRecipe(recipe.id, { reviewBody: this.state.reviewBody });
    this.setState({ reviewBody: '' });
  }

  /**
   * @description - Toggle Favorite Recipe
   *
   * @method toggleFavouriteRecipe
   *
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  toggleFavouriteRecipe(event) {
    event.preventDefault();
    if (this.props.isAuthenticated) {
      this.props.toggleFavouriteRecipe(parseInt(event.target.id, 10));
    }
    checkAuth(this.props.isAuthenticated, this.props.history);
  }

  /**
   * @description - Toggle Thumbs Up
   *
   * @method toggleThumbsUpRecipe
   *
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  toggleThumbsUpRecipe(event) {
    event.preventDefault();
    if (this.props.isAuthenticated) {
      this.props.toggleThumbsUpRecipe(event.target.id);
      this.setState({ toggleThumbsUp: !this.state.toggleThumbsUp });
    }
    checkAuth(this.props.isAuthenticated, this.props.history);
  }

  /**
   * @description - Toggle Thumbs Down
   *
   * @method toggleThumbsDownRecipe
   *
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof RecipeDetail
   */
  toggleThumbsDownRecipe(event) {
    event.preventDefault();
    if (this.props.isAuthenticated) {
      this.props.toggleThumbsDownRecipe(event.target.id);
      this.setState({ toggleThumbsUp: !this.state.toggleThumbsUp });
    }
    checkAuth(this.props.isAuthenticated, this.props.history);
  }


  /**
   * @description - Fetch more reviews
   *
   * @method viewMoreReviews
   *
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof RecipeDetail
   *
   */
  viewMoreReviews(event) {
    event.preventDefault();
    const newReviewsPage = this.state.reviewsPage + 1;
    this.setState({
      reviewsLoaded: false
    });
    const limit = 5;
    const offset = newReviewsPage;
    this.props.getRecipeReviews(this.props.recipe.id, limit, offset);
    this.setState({
      reviewsLoaded: true,
      reviewsPage: newReviewsPage
    });
  }

  /**
 * @description - Renders react component
 *
 * @method render
 *
 * @returns {Jsx} Jsx
 * @memberof RecipeDetail
 */
  render() {
    if (this.props.loaded) {
      const { recipe, user } = this.props;
      const isFavorited = this.props.myFavouriteRecipes
        .includes(parseInt(recipe.id, 10));
      const isUpVoted = recipe.upvotes
        .includes(parseInt(user.userId, 10));
      const isDownVoted = recipe.downvotes
        .includes(parseInt(user.userId, 10));
      if (recipe.name) {
        return (
          <div>
            <main style={{ marginTop: 100 }} id="body">

              <div className="container" id="form">
                <h4 style={{ textAlign: 'center' }} id="recipe-detail-name"
                >{recipe.name}
                </h4><br /><br />

                <div className="row">
                  <div className="col-md-6">
                    <div className="box">
                      <div className="circle">
                        <img className ="circle" src={recipe.image} />
                      </div>
                    </div>
                    <div className="container text-center">
                      <RecipeCardActions
                        isDownVoted={isDownVoted}
                        isUpVoted={isUpVoted}
                        isFavorited={isFavorited}
                        user={user}
                        recipe={recipe}
                        toggleFavouriteRecipe={this.toggleFavouriteRecipe}
                        toggleThumbsDownRecipe={this.toggleThumbsDownRecipe}
                        toggleThumbsUpRecipe={this.toggleThumbsUpRecipe}
                        style={{
                          width: '50%',
                          position: 'relative',
                          margin: '0 auto'
                        }}
                      />
                    </div>
                  </div>
                  <div className="col-md-6">
                    <h5>Description</h5>
                    {recipe.description}
                  </div>
                </div>
                <br />

                <div className="row">

                  <div className="col-md-6">
                    <h5>Ingredients</h5>
                    <ul style={{ fontSize: 20 }}>
                      {
                        recipe.ingredients
                          .map((step, index) => <li key={index}>{step}</li>)
                      }
                    </ul>
                  </div>

                  <div className="col-md-6">
                    <h5>Steps</h5>
                    <ul style={{ fontSize: 20 }}>
                      {
                        recipe.steps
                          .map((step, index) => <li key={index}>{step}</li>)
                      }
                    </ul>
                  </div>
                </div>

                <div className="row">

                  <div className="col-md-6">
                    <CreateReview history={this.props.history}
                      recipeId={recipe.id}
                      onChange={this.onChange}
                      reviewBody={this.state.reviewBody}
                      reviewRecipe={this.reviewRecipe}
                    />
                  </div>
                  <div className="col-md-6" />

                </div><br />

                <div>
                  <h6 className="text-center"
                    style={{
                      color: 'orange',
                      margin: '5 0 10 0',
                      fontSize: 16
                    }}
                  >What People Said
                  </h6>
                </div>
                {recipe.reviews
                  .map(review => <Review key={review.id} review={review}/>)}
              </div>
              <div style={{ display: 'flex', justifyContent: 'space-around' }}>
                {

                  this.props.pagination > this.state.reviewsPage ?
                    <button className="auth-button"
                      onClick={this.viewMoreReviews}
                    >
                    View More
                    </button> :
                    ''
                }
                <Link to="/" onClick={(event => {
                  event.preventDefault();
                  $('html, body').animate({
                    scrollTop: $("#body").offset().top
                  }, 1000);
                })}>Go Up</Link>
              </div>
              <Loader loaded={this.state.reviewsLoaded}/>
              <EditModal recipe={recipe} editRecipe={this.props.editRecipe}/>
              <DeleteModal id={recipe.id} onDelete={this.deleteRecipe}/>
            </main>
          </div>
        );
      }
      return (<Exclamation style={{ marginTop: '200px' }}>
        <p className="text-muted text-center">
          Recipe not found
        </p>
      </Exclamation>);
    }
    return (<Loader loaded={this.props.loaded}/>);
  }
}

const propTypes = {
  isAuthenticated: PropTypes.bool,
  myFavouriteRecipes: PropTypes.array,
  user: PropTypes.object,
  pagination: PropTypes.number,
  recipe: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  toggleFavouriteRecipe: PropTypes.func.isRequired,
  toggleThumbsDownRecipe: PropTypes.func.isRequired,
  toggleThumbsUpRecipe: PropTypes.func.isRequired,
  editRecipe: PropTypes.func.isRequired,
  getRecipe: PropTypes.func.isRequired,
  getRecipeReviews: PropTypes.func.isRequired,
  reviewRecipe: PropTypes.func.isRequired,
  deleteRecipe: PropTypes.func.isRequired
};

RecipeDetail.propTypes = propTypes;

/**
 * mapStateToProps
 * @export
 * @param {Object} state
 * @param {Object} props
 *
 * @return {Object} object
 */
export const mapStateToProps = (state, props) => ({
  user: state.auth.user,
  recipe: state.recipes.singleRecipe,
  isAuthenticated: state.auth.isAuthenticated,
  pagination: Number(state.recipes.pagination),
  loaded: state.recipes.loaded,
  myFavouriteRecipes: state.recipes.userFavouritedRecipeId || [],
});

export default connect(mapStateToProps, {
  getRecipe,
  toggleFavouriteRecipe,
  toggleThumbsUpRecipe,
  toggleThumbsDownRecipe,
  getRecipeReviews,
  reviewRecipe,
  editRecipe,
  deleteRecipe
})(RecipeDetail);

