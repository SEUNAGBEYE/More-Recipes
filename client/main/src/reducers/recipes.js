import actionTypes from '../actions/actionTypes';

const {
  GET_RECIPE,
  GET_RECIPES,
  GET_USER_RECIPES,
  GET_USER_FAVOURITED_RECIPES_ID,
  GET_FAVOURITED_RECIPES,
  GET_POPULAR_RECIPES,
  GET_RECIPE_REVIEWS,
  RECIPE_CATEGORIES,
  SEARCH_RECIPES,
  ADD_RECIPE,
  DELETE_RECIPE,
  EDIT_RECIPE,
  TOGGLE_FAVOURITE_RECIPE,
  TOGGLE_THUMBS_UP_RECIPE,
  TOGGLE_THUMBS_DOWN_RECIPE,
  REVIEW_RECIPE
} = actionTypes;

export const initialState = {
  allRecipes: [],
  singleRecipe: {
    id: '',
    name: '',
    image: '',
    categoryId: '',
    userId: '',
    upvotes: [],
    downvotes: [],
    views: [],
    description: '',
    reviews: []
  },
  pagination: '',
  userFavouritedRecipeId: [],
  favouriteRecipes: [],
  popularRecipes: [],
  recipeCategories: [],
  loaded: false
};

export default (state = initialState, action = {}) => {
  const singleRecipe = { ...state.singleRecipe };
  let allRecipes = [...state.allRecipes];
  const { recipe: recipeOnAction } = action;

  switch (action.type) {
  case ADD_RECIPE:
    return {
      ...state,
      ...{ allRecipes: [action.recipe, ...state.allRecipes] }
    };

  case GET_RECIPE:
    return {
      ...state,
      ...{
        singleRecipe: action.recipe,
        pagination: action.pagination,
        loaded: true
      }
    };

  case GET_USER_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: action.recipes,
        pagination: action.pagination
      }
    };

  case GET_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: action.recipes,
        pagination: action.pagination,
        loaded: true
      }
    };

  case RECIPE_CATEGORIES:
    return {
      ...state,
      ...{
        recipeCategories: action.recipeCategories,
        loaded: true
      }
    };
  case GET_POPULAR_RECIPES:
    return {
      ...state,
      ...{
        popularRecipes: [...action.popularRecipe],
        loaded: true
      }
    };

  case SEARCH_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: [...action.recipes],
        pagination: action.pagination
      }
    };

  case DELETE_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: state.allRecipes
          .filter(recipe => parseInt(recipe.id, 10) !== parseInt(action.id, 10))
      }
    };


  case EDIT_RECIPE:
    if (allRecipes.length >= 1) {
      allRecipes = allRecipes.map((recipe) => {
        if (parseInt(recipe.id, 10) === parseInt(action.recipe.id, 10)) {
          return action.recipe;
        }
        return recipe;
      });
    }

    return {
      ...state,
      ...{
        allRecipes,
        singleRecipe: { ...singleRecipe, ...recipeOnAction }
      }
    };

  case TOGGLE_FAVOURITE_RECIPE:
    return {
      ...state,
      ...{
        favouriteRecipes: !state.userFavouritedRecipeId
          .includes(action.favouritedRecipe.id) ?
          [...state.favouriteRecipes, action.favouritedRecipe] :
          state.favouriteRecipes
            .filter(recipe => recipe.id != action.favouritedRecipe.id),

        userFavouritedRecipeId: !state.userFavouritedRecipeId
          .includes(action.favouritedRecipe.id) ?
          state.userFavouritedRecipeId.concat(action.favouritedRecipe.id) :
          state.userFavouritedRecipeId
            .filter(id => id !== action.favouritedRecipe.id),
        singleRecipe: action.favouritedRecipe
      }
    };

  case TOGGLE_THUMBS_UP_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        singleRecipe: action.recipe
      }
    };

  case TOGGLE_THUMBS_DOWN_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        singleRecipe: action.recipe
      }
    };

  case GET_FAVOURITED_RECIPES:
    return {
      ...state,
      favouriteRecipes: [...action.favouriteRecipes],
      pagination: action.pagination
    };

  case GET_USER_FAVOURITED_RECIPES_ID:
    return {
      ...state,
      ...{ userFavouritedRecipeId: action.favouritedRecipesIds }
    };
  case REVIEW_RECIPE:
    singleRecipe.reviews = [action.reviewBody, ...singleRecipe.reviews];
    return {
      ...state,
      ...{ singleRecipe }
    };
  case GET_RECIPE_REVIEWS:
    singleRecipe.reviews = [...singleRecipe.reviews, ...action.reviews];
    return {
      ...state,
      ...{
        singleRecipe,
        pagination: action.pagination
      }
    };

  default: return {
    ...state,
    ...{ loaded: true }
  };
  }
};

export { actionTypes };
