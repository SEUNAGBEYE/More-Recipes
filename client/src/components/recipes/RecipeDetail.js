import React, { Component } from 'react';
import Navbar from '../common/Navbar';
import Footer from '../common/Footer';
import { pasta, seun } from '../../helpers/Images';

/**
 * @class RecipeDetail
 */ 
export default class RecipeDetail extends Component {

  /**
   * @memberOf RecipeDetail
   * return {object} object
   */
  render(){
    return (
      <div>
        <Navbar />
        <main style={{marginTop: 40}}>
	
          <ul className="breadcrumb" style={{backgroundColor: '#f8f9fa', marginTop: 10}}>
            <li className="breadcrumb-item"><a href="index.html">Home</a></li>
            <li className="breadcrumb-item"><a href="category.html">Category</a></li>
            <li className="breadcrumb-item active">Dessert</li>
          </ul>

          <div className="container">
            <h4 style={{color: 'orange', textAalign: 'center', marginTop: 50}}>Pasta</h4><br /><br />

            <div>
              <div className="box">
                <div className="circle"><img className ="circle" src={pasta} /></div>
              </div>
              <div className="text-center">
                <a href="" className="fa fa-star icons"></a>
                  <a href="" className="fa fa-star icons"></a>
                  <a href="#" className="fa fa-star icons"></a>
                  <a href="#" className="fa fa-star icons"></a>
              </div>
            </div>
            <br />
            <div className="container-fluid">
                <a href="" className="fa fa-eye icons">1000</a>
                  <a href="" className="fa fa-thumbs-up icons">200</a>
                  <a href="#" className="fa fa-thumbs-down icons">10</a>
            </div>
            
            <div className="jumbotron" style={{backgroundColor: '#f8f9fa'}}>
              <h5>Short Description</h5>
              Making pasta at home is much easier than most people would imagine. By making your own pasta dough, you have an unlimited number of options available to you to create your own personal pasta dishes. Imagine delicate layers of egg pasta nestled between a spicy tomato sauce and meltingly tender cheese for unforgettable lasagna. Or maybe you do prefer soft pillows of ravioli stuffed with a tasty ricotta and spinach filling. You too can prepare dishes like this, and many others once you learn the basic technique of making your own pasta.
            </div>

            <div className="jumbotron" style={{backgroundColor: '#f8f9fa'}}>
              <h5>Ingredients</h5>
              <ul style={{fontSize: 20}}>
                <li>flour</li>
                <li>Sugar</li>
                <li>Egg</li>
                <li>Water</li>
                <li>Salt</li>
                <li>Oil</li>
                <li>Dough</li>
              </ul>
            </div>

            <div className="jumbotron" style={{backgroundColor: '#f8f9fa'}}>
              <h5>Steps</h5>
              <ul style={{fontSize: 20}}>
                <li>
                  Egg pasta is super simple ingredient wise: flour, salt, eggs and olive oil. Begin by whisking together 2 cups of flour and 1/2 teaspoon of salt in a large bowl. Make a well in the center of the flour and add three large eggs and one tablespoon of extra virgin olive oil.
                </li>
                <li>						
                  Whisk the eggs and oil with a fork while slowly incorporating some of the flour from around the edges as you go. Once the dough becomes too thick to whisk with a fork, turn the dough out onto a clean work surface, along with any leftover flour from the bowl. Knead the dough and remaining flour until youâ€™ve got a smooth, stiff ball of dough.
                </li>
                <li>
                  Wrap the dough loosely with plastic and let it rest at room temperature for 30 minutes. The wrapped dough can also be stored in the refrigerator for up to a day if necessary.
                </li>
                <li>
                  When you're ready to make your pasta, divide the dough into four smaller portions. Work with one portion at a time, leaving the remaining three covered with plastic to prevent them from drying out.
                </li>
                <li>
                  Dust the dough liberally with flour, then flatten it into a rectangular shape with your hands. Once flattened, begin to roll the dough into a long, thin sheet, dusting liberally with flour on both sides as you go.
                </li>
                <li>
                  Roll the pasta as thin as possible, keeping in mind that it will plump considerably when cooked.
                </li>
                <li>
                  Dust the pasta liberally with flour once again, then fold it into a wide, flat roll. Folding the pasta into a flat roll instead of simply rolling the pasta up into a tube will help prevent the pasta from being squashed as it's cut. Use a sharp knife to cut the pasta into strips.
                </li>
                <li>
                  Shake the cut pasta strands out into a loose pile and dust with flour. At this point, the pasta can be dropped into boiling water to cook or can be piled onto a baking sheet in single portions and frozen.

                  If freezing, transfer the individual frozen pasta nests to an air-tight freezer bag for storage. To dry the frozen pasta, hang the strands through a hanger over a clothes drying rack, or over the back of a chair in a cool, dry place. Once the pasta is dry and brittle, it can be stored in an air-tight container. 
                </li>
                <li>
                  Keep in mind that fresh or fresh-frozen pasta cooks much faster than dried pasta. A quick three- to four-minute boil in lightly salted water is all you need for a plateful of springy, flavorful homemade pasta. The sky's the limit with shapes and flavors, so take this base recipe and make it your own.
                </li>
              </ul>
            </div>

            <div className="row">
              
              <div className="col-md-6">
                
                  <form className="form-group">
                    
                    <textarea rows="5" cols="20" className="form-control" placeholder="Write reviews here"></textarea>
                    <button className="btn-default auth-button" style={{margin: 10, float: 'left'}}> Submit</button>
                  </form>
                          
              </div>
              <div className="col-md-6"></div>

            </div><br />

            <div>
              <h6 style={{color: 'orange', margin: '5 0 10 0', fontSize: 16}} className="text-center">What People Said</h6>
            </div>

            <div className="row" style={{marginBottom: 15}}>
              
              <div className="" style={{backgroundColor: '#f8f9fa', border:'1 solid #f8f9fa'}}>
                
                <div className="col-md-1 review" style={{paddingTop: 10}}>
                <img src={seun} style={{width: 30, height: 30, borderRadius: 80}} />
                </div>
                  
                <div className="col-md-11 review">
                  <h5 style={{color: 'orange', marginTop: 5}}>Joe</h5>

                  Do you have a recipe for an eggless homemade pasta? I have a vegan daughter and would like to make some homemade pasta. I have tried a plain flour and water recipe, but did not like the texture and it cooked up too soft. Thanks Vicki DiFederico
                </div>	
              </div>
            </div>

            <div className="row">
              
              <div className="" style={{backgroundColor: '#f8f9fa', border:'1 solid #f8f9fa'}}>
                
                <div className="col-md-1 review" style={{paddingTop: 10}}>
                <img src={seun} style={{width: 30, height: 30, borderRadius: 80}} />
                </div>
                <div className="col-md-11 review">
                  <h5 style={{color: 'orange', marginTop: 5}}>Joe</h5>

                  Do you have a recipe for an eggless homemade pasta? I have a vegan daughter and would like to make some homemade pasta. I have tried a plain flour and water recipe, but did not like the texture and it cooked up too soft. Thanks Vicki DiFederico
                </div>							
              </div>
            </div>
          </div>
        </main>
        <Footer style={{marginTop: 50}}/>
      </div>
    )
  }
}