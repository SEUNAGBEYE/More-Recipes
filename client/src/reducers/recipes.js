
import isEmpty from 'lodash/isEmpty';

const initialState = {
  allRecipes: [],
  recipesCount: '',
  userFavouritedRecipeId: [],
  favouriteRecipes: [],
  popularRecipes: []
  // favouritedRecipesCount: '',
  // userRecipes: []
};

export default (state = initialState, action = {}) => {
  switch (action.type) {
  case 'ADD_RECIPE':
    return [...state, action.recipe];

  case 'GET_RECIPE':
    return { ...state, ...{ allRecipes: [...state.allRecipes.filter(recipe => recipe.id === parseInt(action.id))] || action.recipe } };

  case 'GET_USER_RECIPES':
    return {
      ...state,
      ...{
        allRecipes:
        action.recipes
      }
    };

  case 'GET_RECIPES':

    return {
      ...state,
      ...{
        allRecipes: action.allRecipes,
        recipesCount: action.recipesCount
      }
    };
  case 'GET_POPULAR_RECIPES':
    console.log('popular', action.popularRecipe);
    return {
      ...state,
      ...{
        popularRecipes: [...action.popularRecipe]
      }
    };

  case 'DELETE_RECIPE':
    return state.allRecipe.filter(recipe => recipe.id !== action.id);


  case 'EDIT_RECIPE':

    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id == action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })]
      }
    };

  case 'TOGGLE_FAVOURITE_RECIPE':
    return {
      ...state,
      ...{
        favouriteRecipes: !state.userFavouritedRecipeId.includes(action.favouritedRecipe.id)
          ? [...state.favouriteRecipes, action.favouritedRecipe]
          : state.favouriteRecipes.filter(recipe => recipe.id != action.favouritedRecipe.id),

        userFavouritedRecipeId: !state.userFavouritedRecipeId.includes(action.favouritedRecipe.id)
          ? state.userFavouritedRecipeId.concat(action.favouritedRecipe.id)
          : state.userFavouritedRecipeId.filter(id => id !== action.favouritedRecipe.id)
      }
    };

  case 'TOGGLE_THUMBS_UP_RECIPE':


    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
      }
    };

  case 'TOGGLE_THUMBS_DOWN_RECIPE':

    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
      }
    };

  case 'GET_FAVOURITED_RECIPES':
    return {
      ...state,
      favouriteRecipes: [...action.favouriteRecipes],
      recipesCount: action.favouritedRecipesCount
    };

  case 'GET_USER_FAVOURITED_RECIPES_ID':
    return {
      ...state,
      ...{ userFavouritedRecipeId: action.favouritedRecipesIds }
    };

  default: return state;
  }
};
