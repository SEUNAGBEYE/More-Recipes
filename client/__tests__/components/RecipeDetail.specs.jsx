
import React from 'react';
import { shallow, mount } from 'enzyme';
import { RecipeDetail, mapStateToProps } from '../../main/src/components/recipes/RecipeDetail';

const state = {
  auth: {
    user: {
      userId: 1,
      firstName: 'Seun',
      lastName: 'Agbeye'
    },
    isAuthenticated: true
  },
  recipes: {
    userFavouritedRecipeId: [1,2,4],
    singleRecipe: recipe
  }
};

const recipe = {
  name: 'Amala',
  image: 'This is the image',
  id: 1,
  upvotes: [],
  downvotes: [],
  ingredients: [],
  steps: [],
  reviews:[]
}

const props = {
  recipe,
  user: state.auth.user,
  myFavouriteRecipes: [],
  reviewRecipe: jest.fn(),
  toggleFavouriteRecipe: jest.fn(),
  toggleThumbsDownRecipe: jest.fn(),
  toggleThumbsUpRecipe: jest.fn(),
  editRecipe: jest.fn(),
  getRecipeReviews: jest.fn(),
  reviewBody: 'This is my review',
  getRecipe: (() =>new Promise((resolve, reject) => {
    recipe ? resolve(recipe) : reject('Recipe not found');
  })),
  deleteRecipe: (() =>new Promise((resolve, reject) => {
    recipe ? resolve(recipe) : reject('Recipe not found');
  })),
  isAuthenticated: state.auth.isAuthenticated,
  history: {
    push: (() => {})
  },
  match: {
    params: {
      id: 1
    }
  },
  user: {
    firstName: 'Seun',
    lastName: 'Agbeye'
  }
};

const event = {
  preventDefault: jest.fn(),
  target: {
    name: 'reviewBody',
    value: 'This is my review',
    id: 1
  }
}

describe('# RecipeDetail', () => {
  const wrapper = shallow(
      <RecipeDetail {...props} />
  );

  it('should render successfully', (done) => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('should call onChange', (done) => {
    const onChangeSpy = jest.spyOn(wrapper.instance(), 'onChange')
    wrapper.instance().onChange(event);
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
    done();
  });
  
  it('should call reviewRecipe', (done) => {
    const reviewRecipeSpy = jest.spyOn(wrapper.instance(), 'reviewRecipe')
    wrapper.instance().reviewRecipe(event);
    expect(reviewRecipeSpy).toHaveBeenCalledTimes(1);
    done();
  });
  it('should call deleteRecipe', (done) => {
    const deleteRecipeSpy = jest.spyOn(wrapper.instance(), 'deleteRecipe')
    wrapper.instance().deleteRecipe(event);
    expect(deleteRecipeSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call toggleFavouriteRecipe', (done) => {
    const toggleFavouriteRecipeSpy = jest.spyOn(wrapper.instance(), 'toggleFavouriteRecipe')
    wrapper.instance().toggleFavouriteRecipe(event);
    expect(toggleFavouriteRecipeSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call toggleThumbsUpRecipe', (done) => {
    const toggleThumbsUpRecipeSpy = jest.spyOn(wrapper.instance(), 'toggleThumbsUpRecipe')
    wrapper.instance().toggleThumbsUpRecipe(event);
    expect(toggleThumbsUpRecipeSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call toggleThumbsDownRecipe', (done) => {
    const toggleThumbsDownRecipeSpy = jest.spyOn(wrapper.instance(), 'toggleThumbsDownRecipe')
    wrapper.instance().toggleThumbsDownRecipe(event);
    expect(toggleThumbsDownRecipeSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call viewMoreReviews', (done) => {
    const viewMoreReviewsSpy = jest.spyOn(wrapper.instance(), 'viewMoreReviews')
    wrapper.instance().viewMoreReviews(event);
    expect(viewMoreReviewsSpy).toHaveBeenCalledTimes(1);
    done();
  });

  it('should call mapStateToProps(state)', (done) => {
    mapStateToProps(state);
    done();
  }); 
});
