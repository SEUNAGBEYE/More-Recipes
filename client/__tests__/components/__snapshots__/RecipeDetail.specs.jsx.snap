// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# RecipeDetail should render successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecipeDetail
    deleteRecipe={[Function]}
    editRecipe={[MockFunction]}
    getRecipe={[Function]}
    getRecipeReviews={[MockFunction]}
    history={
      Object {
        "push": [Function],
      }
    }
    isAuthenticated={true}
    match={
      Object {
        "params": Object {
          "id": 1,
        },
      }
    }
    myFavouriteRecipes={Array []}
    recipe={
      Object {
        "downvotes": Array [],
        "id": 1,
        "image": "This is the image",
        "ingredients": Array [],
        "name": "Amala",
        "reviews": Array [],
        "steps": Array [],
        "upvotes": Array [],
      }
    }
    reviewBody="This is my review"
    reviewRecipe={[MockFunction]}
    toggleFavouriteRecipe={[MockFunction]}
    toggleThumbsDownRecipe={[MockFunction]}
    toggleThumbsUpRecipe={[MockFunction]}
    user={
      Object {
        "firstName": "Seun",
        "lastName": "Agbeye",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "div",
      "loaded": undefined,
      "loadedClassName": "loadedContent",
      "parentClassName": "loader",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "div",
        "loaded": undefined,
        "loadedClassName": "loadedContent",
        "parentClassName": "loader",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
